File: Hard/Next Happy Number/next-happy-number.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG
{
    public static void main(String args[])throws IOException
    {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while(t-- > 0)
        {
            int N = Integer.parseInt(read.readLine());
            Solution ob = new Solution();
            System.out.println(ob.nextHappy(N));
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution{

    static int fun(int n){
        int sum = 0;
        while(n>0){
            int d = n%10;
            sum += d*d;
            n/=10;
        }
        if(sum <= 9){
            return sum;
        }
        else return fun(sum);
    }
    
    static int nextHappy(int N){
        // code here
        int t = N+1;
        while(fun(t) != 1 && fun(t) != 7){
            t++;
        }
        return t;
    }
}

File: Next Happy Number - GFG/next-happy-number.py
#User function Template for python3


class Solution:
    def nextHappy (self, N):
        # code here
        def fun(n):
            s = 0
            while n:
                d = n%10
                s += (d*d)
                n //=10;
            if s <= 9:
                return s
            else:
                return fun(s)
        
        t = N+1
        while (fun(t) != 1) and (fun(t) != 7):
            t += 1
        return t;


#{ 
 # Driver Code Starts
#Initial Template for Python 3


if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        N = int(input())

        ob = Solution()
        print(ob.nextHappy(N))
# } Driver Code Ends